### PR TITLE
Intel CAVS SMP fixes and improvements

### PR DESCRIPTION
**Pull request by lyakh**
Still very much a draft but it's slowly becoming less chaotic

### CODE REVIEW

File: drivers/ipm/ipm_cavs_idc.c
Comment (andyross) [inline drivers/ipm/ipm_cavs_idc.c:232]: I have a half-finished patch that moves the IDC handling into the main SOC code and out of a driver (leaving IPM implemented in terms of that SOC API).    That seems like a cleaner path for this sort of problem.  No reason not to merge this, but IMHO there's a better way long term.

File: include/sys/arch_interface.h
Comment (andyross) [inline include/sys/arch_interface.h:225]: Would suggest something like "arch_cpu_active()".  Even on CAVS, there is a distinction between "powered up" (a bit set in the ADSPCS host register) and "active" (i.e. running Zephyr code and not in ROM or OS initialization, which is what you surely want to know).

File: soc/xtensa/intel_adsp/common/soc_mp.c
Comment (andyross) [inline soc/xtensa/intel_adsp/common/soc_mp.c:208]: Not sure all this is needed, it looks like an attempt to chase races with ROM initialization?   In theory, we shouldn't ever enter this code unless with know with 100% certainty that the target CPU has power and is spinning in the ROM initialization loop waiting on an IDC interrupt.  If we don't know that, there's really no way to do this reliably anyway.

My guess is that this all goes away with a proper host/DSP power management protocol.

**Review by andyross — COMMENTED** [review]
The changes to SMP startup collide pretty badly with the code I'm about to submit for power management (which is a different paradigm that requires host driver changes), but then we always knew that was likely.  This will no doubt go through a few rounds of review anyway.

File: kernel/smp.c
Comment (nashif) [inline kernel/smp.c:85]: lets not have CONFIG_SOF in here. Instead, I would define a new general Kconfig, for example CONFIG_SMP_START_ONE or whatever and make CONFIG_SOF select that...

File: include/arch/xtensa/cache.h
Comment (lyakh) [reply include/arch/xtensa/cache.h:34]: @andyross search for `XCHAL_ERRATUM_497` in SOF sources, e.g. https://github.com/thesofproject/sof/blob/5aeb29b5546e43a8303c3386cc1fc0c4673bb380/src/arch/xtensa/include/xtensa/cacheasm.h#L488 and then https://github.com/thesofproject/sof/blob/5aeb29b5546e43a8303c3386cc1fc0c4673bb380/src/arch/xtensa/include/xtensa/cacheasm.h#L644 and https://github.com/thesofproject/sof/blob/5aeb29b5546e43a8303c3386cc1fc0c4673bb380/src/arch/xtensa/include/xtensa/cacheasm.h#L660 but no, I haven't hit the problem itself

File: kernel/include/kswap.h
Comment (lyakh) [reply kernel/include/kswap.h:109]: Sure, let's improve this. The whole dummy thread story deserves fixing. With a bit of tweaking I've moved this code to SOF SMP start-up, so, no changes here are needed.

File: include/sys/arch_interface.h
Comment (lyakh) [reply include/sys/arch_interface.h:225]: ack

File: kernel/smp.c
Comment (lyakh) [reply kernel/smp.c]: @nashif how about `CONFIG_SMP_BOOT_DELAY`?

File: soc/xtensa/intel_adsp/common/soc_mp.c
Comment (lyakh) [reply soc/xtensa/intel_adsp/common/soc_mp.c]: @andyross Some of that indeed isn't needed, I removed it. But some changes I haven't been able to get rid of. Need to try a bit harder or see if this also works for other Xtensa SMP tests

File: soc/xtensa/intel_adsp/cavs_v15/Kconfig.defconfig.series
Comment (nashif) [inline soc/xtensa/intel_adsp/cavs_v15/Kconfig.defconfig.series:71]: this is using an SOF Kconfig that is duplicated below, we need to fix that. Maybe in SOF itself set CORE_COUNT to what Zephyr provide using MP_NUM_CPUS

File: include/arch/xtensa/cache.h
Comment (andyross) [reply include/arch/xtensa/cache.h:34]: Would still like to see this in some kind of #ifdef and limited to systems that actually care, even if it's all the LX CPUs or whatever.  Obviously not a big deal performance-wise, but without a way to resolve what "Erratum 497" means (it's a Cadence-internal number for something maintainers can't look up) this really amounts to just voodoo.

File: kernel/init.c
Comment (andyross) [reply kernel/init.c:223]: Again, I'd prefer to leave out code that doesn't do anything.  Please remove.

File: kernel/Kconfig
Comment (andyross) [inline kernel/Kconfig:795]: Still don't love this, but it's basically harmless.  We can come in and refactor the way SMP CPU bringup happens later I guess.

File: kernel/sched.c
Comment (andyross) [inline kernel/sched.c:311]: Not sure I buy the aesthetics here.  The point to having that in a separate block at the top level is to clarify that it's a separate (and really important) operation where the current thread is being re-added to the run queue having been preempted.  Obviously the compiler is going to generate the same code either way, but honestly I like this the way it is.   Please remove.

File: arch/xtensa/include/kernel_arch_func.h
Comment (andyross) [inline arch/xtensa/include/kernel_arch_func.h:78]: Yes it does, and that's documented in the Kconfig where it happens.   In general there's no need to call out "Hey!  I'm using an optional API!", you just use it.  Don't think this comment is needed, but not a big deal.

File: arch/xtensa/include/kernel_arch_func.h
Comment (andyross) [inline arch/xtensa/include/kernel_arch_func.h:115]: Not understanding this logic.  The idea of a DUMMY thread is that it has no stack at all, we're running with no current thread, and on the interrupt stack.  The interrupt stack is always local to a specific CPU and thus never needs to be flushed or invalidated.  Is this fix really required?  If so, I think the root problem must be somewhere else.

File: lib/os/p4wq.c
Comment (andyross) [inline lib/os/p4wq.c:273]: Just pick one.  There's no reason not to use APPLICATION all the time, this doesn't have any significant dependancies we need to wait for.

Though... we should probably be defining the SMP initlevel anyway just for compatibility, to avoid code like this...

File: include/sys/p4wq.h
Comment (andyross) [inline include/sys/p4wq.h:36]: One word isn't really much to complain about, but this is a paradigm collision.  There's no reason the application code can't put exactly the same pointer into their own struct.

Not worth a -1, but I don't like this and we should only have one API style for how to find the "owner" of a work item.  Other data structures in Zephyr don't do this.

File: lib/os/p4wq.c
Comment (andyross) [inline lib/os/p4wq.c:99]: Really not liking this.  You can't take a blocking primitive in the main loop, it will stall the whole queue.  Consider a semaphore if you need to have user threads waiting, that can be given without blocking.

**Review by andyross — CHANGES_REQUESTED** [review]
Sorry, I came back to add a note to a comment I'd left on Monday and realized I never actually submitted the review.

A few things that would be really good to fix before merge, but this is a fairly soft -1 and we can do the cleanup in the tree too.

File: include/arch/xtensa/cache.h
Comment (lyakh) [reply include/arch/xtensa/cache.h:34]: ok, let's drop it for now and try to find out more later

File: arch/xtensa/include/kernel_arch_func.h
Comment (lyakh) [reply arch/xtensa/include/kernel_arch_func.h:115]: This is needed because if `arch_cohere_stacks()` returns there, it doesn't set `EXCSAVE3`, then in `xtensa_switch` in arch/xtensa/core/xtensa-asm2-util.S it reads 0 from `EXCSAVE3` and flushes a huge address range, which also delays the secondary core start up.

File: lib/os/p4wq.c
Comment (lyakh) [reply lib/os/p4wq.c:99]: in principle I understand your argument, but have a look what it's racing against. `k_p4wq_wait()` below really cannot hold the mutex for any significant time. And it's a per-work item mutex. So, I don't think this can create any real congestion. As for the semaphore - sure, in principle I could do that, but a semaphore is a counting primitive, which isn't what we need here.
Besides, look at `k_p4wq_wait()` again: if `.sync` isn't set and `k_p4wq_wait()` is called to regain ownership of the work item, without the mutex we have a race there - it can be called right after the handler resubmitted the work item and before we re-take the spin-lock. In that case `k_p4wq_wait()` can wrongly return 0, i.e. "work item belongs to us" and the caller will free it too early.

File: arch/xtensa/include/kernel_arch_func.h
Comment (andyross) [reply arch/xtensa/include/kernel_arch_func.h:115]: Right, but that's the protection the original code offered.  It ddn't flush dummy stacks.  The new code **will** try to flush a dummy stack, but will also refuse to flush a stack with a zero size.  Now... it's probably true that the stack size field of a dummy stack is zero (but AFAIK nothing in the architecture guarantees that, dummy thread objects exist as temporary stack data and I frankly don't know if they're fully initialized or not), so... maybe this is a noop.

But if it's a noop we should just skip it.  And if it's not we should be clear about exactly what behavior is desired.  To my eyes the original test is clearer.

Comment (lyakh) [reply arch/xtensa/include/kernel_arch_func.h:115]: @andyross the problem is that it is very much *not* a NOP. And that is a bug. The fact, that secondary core takes 0 from `EXCSAVE3`, subtracts some valid address from it (or the other way round) and runs cache flushing on all those megabytes is a bug IMHO. Maybe my fix isn't perfect or even not correct. But I'd really like to fix this. The other option is in `xtensa_switch` to check `EXCSAVE3` for 0 and skip flushing in such a case? Would that be a correct solution?

Comment (andyross) [reply arch/xtensa/include/kernel_arch_func.h:114]: I'm still not understanding why you're getting a zero in that field for a non-dummy thread.  Something else is wrong if that's happening.  I just tried with a quick __ASSERT and didn't see this, how are you making it happen?

File: lib/os/p4wq.c
Comment (andyross) [reply lib/os/p4wq.c:99]: I'll look closer to see exactly what this is trying to protect, but "no blocking primitives in a work queue main loop" is a pretty firm design rule here.  Again, I'm 99% certain that the correct IPC primitive for what you want is a semaphore (which allows another thread to block, but which doesn't ever block the thread calling k_sem_give()) and not a priority-inheriting mutex.

File: arch/xtensa/include/kernel_arch_func.h
Comment (lyakh) [reply arch/xtensa/include/kernel_arch_func.h:114]: @andyross sorry for not explaining properly. It **is** the dummy thread, running on a secondary core, using the IRQ stack.

Comment (lyakh) [reply arch/xtensa/include/kernel_arch_func.h:114]: So what this change does it allows the dummy thread to proceed and set `EXCSAVE3` accordingly instead of bailing out on the `if` above

File: lib/os/p4wq.c
Comment (lyakh) [reply lib/os/p4wq.c:99]: @andyross ok, I think this race can also be avoided by just checking the semaphore counter, using `k_sem_count_get()` in the non-blocking ownership re-acquisition case. We would just take the semaphore in `k_p4wq_submit()` and give it in `p4wq_loop()` right at the end of the `if (r)` block. I think your `thread_was_requeued()` trick wouldn't be needed then either.

**Comment by lyakh** [comment]
CI breakages can be fixed by https://github.com/thesofproject/sof/pull/4094

File: soc/xtensa/intel_adsp/cavs_v15/Kconfig.defconfig.series
Comment (lgirdwood) [reply soc/xtensa/intel_adsp/cavs_v15/Kconfig.defconfig.series]: @lyakh @marc-hb please align fully with Zephyr on all Kconfig symbols - i.e. SOF will follow the Zephyr naming/usage.

**Comment by lgirdwood** [comment]
@lyakh does this need to wait for @andyross PM work or can we cleanup/rebase some smaller parts and upstream in smaller chunks ?

**Comment by lyakh** [comment]
> @lyakh does this need to wait for @andyross PM work or can we cleanup/rebase some smaller parts and upstream in smaller chunks ?

@lgirdwood no, this doesn't have to wait for anything. @nashif asked on Monday for this to be merged this week... During the same conversation it also was mentioned, that splitting this into multiple PRs wasn't favoured.
One issue that is blocking this PR are compilation breakages related to various unrelated regressions. To fix those @nashif could update to the latest SOF "main" branch or at least pull https://github.com/thesofproject/sof/pull/4094, https://github.com/thesofproject/sof/pull/4100, https://github.com/thesofproject/sof/pull/4078. But most of all this PR has to be reviewed by @andyross again.

**Comment by lyakh** [comment]
@andyross @nashif ping, rebased on top of current "master," fixed a typo, no changes otherwise

**Comment by lyakh** [comment]
I've just pushed https://github.com/lyakh/zephyr/pull/1 which for now I've based on top of my branch, containing these commits. Once this PR is merged, I'll rework the new one to the upstream tree too.

**Review by andyross — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**
